### DISCUSSION TITLE
开发相关问题

### DISCUSSION THREAD

**Discussion by DryIce-cc** [ClassIsland/ClassIsland#200] [Q&A]
在 Windows 10 上使用 Rider 对 ClassIsland 进行开发；没有 C# 和 WPF 经验，没用过 VS Code 以外的编辑器，会基本的 Google 搜索。
来此询问各位大佬经验，感激不尽。

1. (如图) Margin 边距设置是否有一定规范？是否有直观的界面编写器？自带的预览不正确(如图)应当如何解决？
![image](https://github.com/ClassIsland/ClassIsland/assets/165131008/8ae18609-e1e8-4128-b611-e86a3c6de62e)

2. 构建并运行 ClassIsland (Quiet) 后，如何使用现成的配置文件？
（如果重启应用拷配置文件，rider 的 debug 会掉；看到配置里有个 ClassIsland: FolderProfile，但似乎关系不大）

3. 每次测试改动，只能等待良久其重新构建和启动吗？有没有更快的方法？

**Comment by Doctor-yoi** [comment]
1.本项目逻辑复杂，预览经常无法解析或预览出错，建议使用人脑渲染的方式进行预览
2.你直接给生成的配置文件替换了不就得了😧
3.理论上可以热重载，但是我觉得挺难蚌的

**Reply by DryIce-cc** [reply]
谢谢🙏我和你的做法一样，但这样就没法在rider内debug了，且需要重启ci

**Comment by LiuYan-xwx** [comment]
margin顺序是 左 上 右 下
[FrameworkElement.Margin 属性](https://learn.microsoft.com/zh-cn/dotnet/api/system.windows.frameworkelement.margin?view=windowsdesktop-8.0)

**Reply by LiuYan-xwx** [reply]
配置文件，档案的话，我是改了个名字，让它使用那个档案就行了。settings我是没遇到会重置

**Reply by DryIce-cc** [reply]
我这里是只要重新构建，原来的bin\Debug\net8.0-windows都会被覆盖掉，一开就是ci设置向导（这玩意还没办法跳过

**Reply by LiuYan-xwx** [reply]
> 我这里是只要重新构建，原来的bin\Debug\net8.0-windows都会被覆盖掉，一开就是ci设置向导（这玩意还没办法跳过

可能是ide的问题吧()我用的vs，我这边不会出现settings.json被覆盖导致每次都进入向导

**Comment by DryIce-cc** [comment]
肥肠抱歉，我不知道一个ask只能设定一个answer，很感谢你们的帮助 @Doctor-yoi 和 @LiuYan-xwx [流萤_比心]

**Comment by DryIce-cc** [comment]
**我注意到 Helpers/TimeSpanFormatHelper.cs 内提供了从 TimeSpan 到“h小时m分钟s秒”的转换**
（如下图 ClassIsland/Services/NotificationProviders/ClassNotificationProvider.cs，上课提醒的语音播报内容）
![image](https://github.com/ClassIsland/ClassIsland/assets/165131008/2577aac4-b977-4597-a77d-ddca32c2935e)

**但是这个 Helper 只在 .cs 文件中使用过，.xaml 文件中都用的是 ClassIsland.Core/Converter 类**
（如下图 ClassIsland/Controls/NotificationProviders/ClassNotificationProviderControl.xaml，上课提醒的显示内容）
![image](https://github.com/ClassIsland/ClassIsland/assets/165131008/e8ee5721-bea3-42cc-a24e-24068b0f0725)

**我想知道怎么在 ClassNotificationProviderControl.xaml 中使用 Helper 类**，
写成 Helper 类的转换器和直接用 Converter 类有什么区别，
以及为什么相同的的内容要用不同语言实现两遍（能否合并他们？），
谢谢。

**Comment by Doctor-yoi** [comment]
建议cv一下搞一个新的converter

获取 Outlook for iOS<https://aka.ms/o0ukef>
________________________________
发件人: DryIce-cc ***@***.***>
发送时间: Friday, July 5, 2024 11:34:08 AM
收件人: ClassIsland/ClassIsland ***@***.***>
抄送: Doctor-yoi ***@***.***>; Mention ***@***.***>
主题: Re: [ClassIsland/ClassIsland] 开发相关问题 (Discussion #200)


我注意到 Helpers/TimeSpanFormatHelper.cs 内提供了从 TimeSpan 到“h小时m分钟s秒”的转换
（如下图 ClassIsland/Services/NotificationProviders/ClassNotificationProvider.cs，上课提醒的语音播报内容）
image.png (view on web)<https://github.com/ClassIsland/ClassIsland/assets/165131008/2577aac4-b977-4597-a77d-ddca32c2935e>

但是这个 Helper 只在 .cs 文件中使用过，.xaml 文件中都用的是 ClassIsland.Core/Converter 类
（如下图 ClassIsland/Controls/NotificationProviders/ClassNotificationProviderControl.xaml，上课提醒的显示内容）
image.png (view on web)<https://github.com/ClassIsland/ClassIsland/assets/165131008/e8ee5721-bea3-42cc-a24e-24068b0f0725>

我想知道怎么在 ClassNotificationProviderControl.xaml 中使用 Helper 类，
写成 Helper 类的转换器和直接用 Converter 类有什么区别，
以及为什么相同的的内容要用不同语言实现两遍（能否合并他们？），
谢谢。

―

**Comment by HelloWRC** [comment]
> (如图) Margin 边距设置是否有一定规范？是否有直观的界面编写器？自带的预览不正确(如图)应当如何解决？

用 vs 吧，rider 对 wpf 支持没 vs 那么好。

此外，建议在开发 ClassIsland 本体时养成不用设计器的习惯，因为很多界面设计器都打不开（）

**Reply by LyCecilion** [reply]
怪不得我的 Visual Studio 设计器打不开，还以为是我自己的问题😢

**Comment by HelloWRC** [comment]
> 每次测试改动，只能等待良久其重新构建和启动吗？有没有更快的方法？

可以用热重载。vs里xaml可以实时热重载，更改完可以马上在应用内看到效果。

**Reply by DryIce-cc** [reply]
这不巧了吗，rider 里是 cs 可以实时热重载（）我试下用 vs

24/7/28 更：vs 实时热重载确实真好用啊（）
